core/commands/files/files.go

```diff
@@ -428,10 +435,328 @@ Examples:
+					fmt.Fprintln(w, output.Hash)
+				} else {
+					fmt.Fprintln(w, item.Name)
+				}
+
+				return nil
+			}
+
+			if output.Type == summaryType {
+				summary := output.Summary
+
+				fmt.Fprintf(w, "%d directories, %d files present\n",
+					summary.NbDir,
+					summary.NbFile,
+				)
+
+				if summary.SizeTotal > 0 {
+					fmt.Fprintf(w, "%s present of %s (%.2f%%)",
+						humanize.Bytes(uint64(summary.SizeLocal)),
+						humanize.Bytes(uint64(summary.SizeTotal)),
+						100.0*float64(summary.SizeLocal)/float64(summary.SizeTotal),
+					)
+				} else {
+					fmt.Fprintf(w, "%s present of unknown", humanize.Bytes(uint64(summary.SizeLocal)))
+				}
+
+				fmt.Fprintln(w)
+				return nil
+			}
+
+			return fmt.Errorf("unknown output type")
+		}),
+	},
+}
+
+type walkState struct {
+	ctx     context.Context
+	res     cmds.ResponseEmitter
+	dagserv dag.DAGService
+	node    node.Node
+
+	onlySummary       bool
+	depth             int
+	name              string
+	parentIsDirectory bool
+	lastChild         bool
+}
+
+func walkBlockStart(ctx context.Context, res cmds.ResponseEmitter, dagserv dag.DAGService, nd node.Node, isDirectory bool, onlySummary bool) (treeSummary, error) {
+	return walkBlock(walkState{
+		ctx,
+		res,
+		dagserv,
+		nd,
+		onlySummary,
+		0,
+		nd.Cid().String(),
+		isDirectory,
+		false,
+	})
+}
+
+// Build a summary of a dag while emiting treeOutput with treeItem to display the graph of files/directories
+func walkBlock(state walkState) (treeSummary, error) {
+
+	// Start with the block data size
+	result := treeSummary{
+		Local:     true,
+		SizeLocal: uint64(len(state.node.RawData())),
+	}
+
+	isDirectory := false
+	// try to decode unixfs
+	if pn, ok := state.node.(*dag.ProtoNode); ok {
+		if unixfs, err := ft.FromBytes(pn.Data()); err == nil {
+
+			// if cumulative size is available, use it as total size
+			cumSize, err := pn.Size()
+			if err == nil {
+				result.SizeTotal = cumSize
+			}
+
+			unixType := unixfs.GetType()
+
+			// To distinguish files from chunk
+			isFile := unixType == ft.TFile && state.parentIsDirectory
+			isDirectory = unixType == ft.TDirectory
+
+			if isFile {
+				result.NbFile++
+			}
+
+			// Don't count the root directory if any
+			if isDirectory && state.depth != 0 {
+				result.NbDir++
+			}
+
+			if !state.onlySummary && (isFile || isDirectory) {
+				state.res.Emit(&treeOutput{
+					Hash: state.node.Cid().String(),
+					Type: itemType,
+					Item: treeItem{
+						Name:      state.name,
+						Depth:     state.depth,
+						LastChild: state.lastChild,
+					},
+				})
+			}
+		}
+	}
+
+	nbChild := len(state.node.Links())
+	for i, link := range state.node.Links() {
+		child, err := state.dagserv.Get(state.ctx, link.Cid)
+
+		isLastChild := i == nbChild-1
+
+		if err == dag.ErrNotFound {
+			result.Local = false
+
+			if !state.onlySummary {
+				state.res.Emit(&treeOutput{
+					Hash: link.Cid.String(),
+					Type: itemType,
+					Item: treeItem{
+						Name:      "[missing]",
```

Comment:
Good example of where this is a reasonable CLI command but a bad API endpoint. This is *not* the name of the file (also, confusing if I named a file `[missing]`).